### PR TITLE
Unify frontend

### PR DESCRIPTION
**Pull request by ftshijt**
## What?

Unify frontend setting

## Why?

- Split rawnet to separate frontend and encoder
- Add s3prl and fbank frontend
- Fix auto-infer of spk number

### CODE REVIEW

**Comment by ftshijt** [comment]
I'm still doing some formating stuff, so please pending merge the PR~ But the general design is finished, so please feel free to leave comments.

File: egs2/voxceleb/spk1/conf/tuning/train_rawnet3_sgdr_bs.yaml [hunk 1]
```diff
@@ -25,33 +32,35 @@ preprocessor_conf:
...
-  nclasses: 7205
   margin: 0.3
   scale: 30
 
+# Training related
 max_epoch: 40
-#num_iters_per_epoch: 1
 num_att_plot: 0
-num_workers: 8
+num_workers: 0
```
Comment (Jungjee) [inline egs2/voxceleb/spk1/conf/tuning/train_rawnet3_sgdr_bs.yaml]: more close to a question. Do we normally set default num_workers to 0 in cfgs?

File: espnet2/spk/layers/ecapa_block.py [hunk 2]
```diff
@@ -23,9 +23,9 @@ def forward(self, input):
         return input * x
 
 
-class Bottle2neck(nn.Module):
+class EcapaBlock(nn.Module):
     def __init__(self, inplanes, planes, kernel_size=None, dilation=None, scale=8):
-        super(Bottle2neck, self).__init__()
+        super(self).__init__()
```
Comment (Jungjee) [inline espnet2/spk/layers/ecapa_block.py]: super().__init__() ?

**Review by Jungjee — APPROVED** [review]
Code looks much better!
Thank you for the work!

File: egs2/voxceleb/spk1/conf/tuning/train_rawnet3_sgdr_bs.yaml (hunk 1 above)
Comment (ftshijt) [reply egs2/voxceleb/spk1/conf/tuning/train_rawnet3_sgdr_bs.yaml]: It depends on the CPU cores attached, but I found that using multiple one would cause issue in collect stats stage. But still, it might not be necessary to keep 0, I will change it back.

File: espnet2/spk/layers/ecapa_block.py (hunk 2 above)
Comment (ftshijt) [reply espnet2/spk/layers/ecapa_block.py]: Oh, I forget the ecapa, will add it

**Comment by ftshijt** [comment]
Hi @Jungjee , I think it is ready to merge if you are satisfied with the current implementation, so that we can combine the discussion in your espnet PR.

File: egs2/voxceleb/spk1/conf/tuning/train_ecapa_tdnn.yaml [hunk 3]
```diff
@@ -1,44 +1,56 @@
...
+# Frontend
+frontend: default
+frontend_conf:
+    hop_length: 128
+    n_mels: 80
 
+# Normalizer
+normalize: utterance_mvn
+normalize_conf:
+    norm_vars: false
```
Comment (ftshijt) [inline egs2/voxceleb/spk1/conf/tuning/train_ecapa_tdnn.yaml:12]: This is an example of adding normalize function with only mean subtraction on utterance level. We would need some modification in the collect stats stage for global mvn, but since we are unlikely to use it, I just add this one here instead of adding tens of lines in the shell script.
